--- ElGamal.py
+++ ElGamal.py
@@ -62,7 +62,11 @@
 
         c1 = pow(g, r, p)
         # new scheme: c2 = pk^r * m (mod p)
-        c2 = (pow(pk, r, p) * (m_int % p)) % p
+        val = (m_int + 1) % p
+        if pow(val, q, p) != 1:
+            val = p - val
+        
+        c2 = (pow(pk, r, p) * val) % p
         return Ciphertext(c1, c2, self.params)
 
     def dec(self, sk: SecretKey, ciphertext: Ciphertext) -> Message:
@@ -70,8 +74,10 @@
         p = self.params.p
         c1, c2 = ciphertext
         x = pow(c1, sk, p)
-        # gm = m_int (mod p)
-        m_int = (c2 * pow(x, -1, p)) % p
+        M = (c2 * pow(x, -1, p)) % p
+        if M > self.params.q:
+            M = p - M
+        m_int = (M - 1) % p
         return m_int
 
 
@@ -99,7 +105,10 @@
         for value in d_i_power_basis:
             d = (d * value) % self.params.p
 
-        m_int = (c2 * pow(d, -1, self.params.p)) % self.params.p
+        M = (c2 * pow(d, -1, self.params.p)) % self.params.p
+        if M > self.params.q:
+            M = self.params.p - M
+        m_int = (M - 1) % self.params.p
         return m_int
 
     def calculate_di_for_shamir(self, c1, share):
